articles/techniques/testing-and-debugging.md

```diff
@@ -157,11 +120,15 @@ After you execute a test in Test Explorer and click on the test, a panel will ap
 #### [Command Line / Visual Studio Code](#tab/tabid-vscode)
 
 The pass/fail status for each test is printed to the console by `dotnet test`.
-For failing tests, the outputs logged as a result of the `output.WriteLine(msg)` call above are also printed to the console to help diagnose the failure.
+For failing tests, the outputs are also printed to the console to help diagnose the failure.
 
 ***
 
-### Assertions
+## Assertions
+
+One important consequence of the fact that functions in Q# have no side effects is that any effects of executing a function whose output type is the empty tuple `()` can never be observed from within a Q# program.
```

Comment:
```suggestion
Because functions in Q# have no _logical_ side effects, any _other kinds_ of effects of executing a function whose output type is the empty tuple `()` can never be observed from within a Q# program.
```